### ISSUE TITLE
BUG: optimizer.compute_gradients() produces inconsistent gradient with the same training instance and label 

### ISSUE THREAD

**Issue by EdwardLin2014** [tensorflow/tensorflow#21515]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu Server & Window 10
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**:
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: v1.8.0-0-g93bc2e2072 & v1.9.0-0-g25c197e023
- **Python version**: 3.6.2
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: V9.0.178
- **GPU model and memory**: NVIDIA Tesla V100-SXM2-16GB & NVIDIA GeForce GTX 1080 Ti 11GB
- **Exact command to reproduce**:
### Source code / logs

```
import tensorflow as tf
import numpy as np

#########################################################################
## Define Tensorflow Wrapper
print('Define Tensorflow Wrapper......')
def weight_variable(shape):
    """Create a bias variable with appropriate initialization."""
    initial = tf.constant(0.2, shape=shape)
    return tf.Variable(initial)

def bias_variable(shape):
    """Create a bias variable with appropriate initialization."""
    initial = tf.constant(0.1, shape=shape)
    return tf.Variable(initial)

def initialize_variable(shape, name):
    with tf.name_scope(name + '_layer'):
        with tf.name_scope('weights'):
            weights = weight_variable(shape)
        with tf.name_scope('biases'):
            biases = bias_variable([1,shape[3],1,1])
        return weights, biases

def conv2dLayer(inputs, weights, biases, name, act=tf.nn.relu):
    preactivate = tf.nn.conv2d(inputs, weights, strides=[1,1,1,1], padding='VALID', data_format="NCHW", dilations=[1,1,1,1], name='conv2d') + biases
    return act(preactivate, name='activation')

def max_pool_3x3(x, name):
    return tf.nn.max_pool(x, ksize=[1, 1, 3, 3], strides=[1, 1, 3, 3], padding='VALID', data_format="NCHW", name=name)
    
#########################################################################
## Create Neural Network
print('Create Neural Network......')
# Initial Trainable Parameters
Conv1W,Conv1b = initialize_variable([3,3,1,2],'Conv1')
Conv2W,Conv2b = initialize_variable([3,3,2,2],'Conv2')
Conv3W,Conv3b = initialize_variable([3,3,2,2],'Conv3')
Conv4W,Conv4b = initialize_variable([3,3,2,2],'Conv4')
fc1W,fc1b = initialize_variable([39,14,2,372],'fc1')
fc2W,fc2b = initialize_variable([1,1,372,372],'fc2')
fc3W,fc3b = initialize_variable([1,1,372,1],'fc3')
# Placeholder for input and output
x = tf.placeholder(tf.float32, [None,1,372,None], name='xInput')
y_ = tf.placeholder(tf.float32, [None,1,1,None], name='yInput')
# Architecture
Conv1 = conv2dLayer(x,Conv1W,Conv1b,'Conv1')
Conv2 = conv2dLayer(Conv1,Conv2W,Conv2b,'Conv2')
MaxPool1 = max_pool_3x3(Conv2,'MaxPool1')
Conv3 = conv2dLayer(MaxPool1,Conv3W,Conv3b,'Conv3')
Conv4 = conv2dLayer(Conv3,Conv4W,Conv4b,'Conv4')
MaxPool2 = max_pool_3x3(Conv4,'MaxPool2')
fc1 = conv2dLayer(MaxPool2,fc1W,fc1b,'fc1')
fc2 = conv2dLayer(fc1,fc2W,fc2b,'fc2')
y = conv2dLayer(fc2,fc3W,fc3b,'fc3',act=tf.identity)
# Define loss
crossEntropy = tf.nn.sigmoid_cross_entropy_with_logits(labels=y_, logits=y, name='crossEntropy')
lossValue = tf.reduce_mean(crossEntropy, name='lossValue')
# Define optimizer
#optimizer = tf.train.AdadeltaOptimizer(learning_rate=0.001,rho=0.95,epsilon=1e-08,use_locking=False,name='Adadelta')
#optimizer = tf.train.AdagradOptimizer(learning_rate=0.01,initial_accumulator_value=0.1,use_locking=False,name='Adagrad')
#optimizer = tf.train.AdamOptimizer(learning_rate=0.001,beta1=0.9,beta2=0.999,epsilon=1e-08,use_locking=False,name='Adam')
#optimizer = tf.train.MomentumOptimizer(learning_rate=0.01, momentum=0.95, name='nestrov', use_nesterov=True)
optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.01, name='GradientDescent')
gvs = optimizer.compute_gradients(lossValue)
train_step = optimizer.apply_gradients(gvs)

#########################################################################
## Illustrate compute_gradients() random issue
print('Load x and y input......')
data = np.load('testGrad.npz')
xInput = data['x']
yInput = data['y']

print("Session 1")
with tf.Session() as sess1:
    sess1.run(tf.global_variables_initializer())
    lossValue_eval1, gvs_eval1 = sess1.run([lossValue,gvs], feed_dict={x:xInput,y_:yInput})
    sess1.run(train_step, feed_dict={x:xInput,y_:yInput})
    lossValue1_eval1, gvs1_eval1, = sess1.run([lossValue,gvs], feed_dict={x:xInput,y_:yInput})
    sess1.close()

print("Session 2")
with tf.Session() as sess2:
    sess2.run(tf.global_variables_initializer())
    lossValue_eval2, gvs_eval2 = sess2.run([lossValue,gvs], feed_dict={x:xInput,y_:yInput})
    sess2.run(train_step, feed_dict={x:xInput,y_:yInput})
    lossValue1_eval2, gvs1_eval2, = sess2.run([lossValue,gvs], feed_dict={x:xInput,y_:yInput})
    sess2.close()

print('--------------------------------------------------')
print('Check whether they have the same values in each session')
print('Before Training: ')
print('lossValue? %r' % all(lossValue_eval1==lossValue_eval2))
check_gvs_BT = np.zeros((14,2))
for i in range(14):
    for j in range(2):
        check_gvs_BT[i][j] = all(gvs_eval1[i][j] == gvs_eval2[i][j])
print('gvs? %r' % all(check_gvs_BT))
print('After Training: ')
print('lossValue? %r' % all(lossValue1_eval1==lossValue1_eval2))
check_gvs_AT = np.zeros((14,2))
for i in range(14):
    for j in range(2):
        check_gvs_AT[i][j] = all(gvs1_eval1[i][j] == gvs1_eval2[i][j])
print('gvs? %r' % all(check_gvs_AT))

tf.reset_default_graph()
```

### Describe the problem
By executing the above code with the fixed training instance and label (which you can download from https://www.dropbox.com/s/828v71z4tm399z2/testGrad.npz?dl=0 ), we can see that the gradient value computed by compute_gradients(),  gvs_eval1[0][0][0,0,0,0] & gvs_eval2[0][0][0,0,0,0] are not always the same in each session. This happens regardless which optimizer you use.

Correct me if I have made some stupid mistake in this code. As I have tried simple network architecture (for example, training y = weight * x), and compute_gradients() seems to be working fine.

**Comment by michaelisard** [comment]
How different are they? GPU computations are nondeterministic and so floating point errors can be different between runs.

**Comment by michaelisard** [comment]
OK I hadn't realized you are reinitializing the weights to new random values for each run correct? I think this is probably the expected behavior.

**Comment by EdwardLin2014** [comment]
? FYI, based on my code, I did NOT randomizes the initial value of the weights. I make them all constant with tf.constant().

Because of the non-deterministic nature of GPU ops, now I know it is the tf.nn.conv2d() cause optimizer.compute_gradients() to compute random gradients. 

By the way, is it possible that, in the tensorflow doc, to list out which tensorflow API/ops are
1. non-deterministic/deterministic
2. going to have an option for choosing non-deterministic/deterministic cuDNN algo
(it seems you guys are working on that :)   really appreciate your hard work :)
https://github.com/tensorflow/tensorflow/issues/12871
https://github.com/tensorflow/tensorflow/pull/10636 )

**Comment by michaelisard** [comment]
Yes sorry I didn't read your code properly.

@rmlarsen do you think it's feasible to maintain a list of deterministic ops in TF?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 89 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @michaelisard: It has been 104 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@EdwardLin2014,
I guess **`Determinism`** in **`Tensorflow GPU`** is possible now using [tensorflow-determinism](https://pypi.org/project/tensorflow-determinism/). Can you PTAL. Thanks!
